mathics/builtin/lists.py

```diff
@@ -2096,14 +2099,235 @@ def apply(self, list, sep, evaluation):
+    # the hash code matches, is SameQ evoked.
+    #
+    # this assumption has been checked for these types: Integer, Real, Complex,
+    # String, Rational (*), Expression, Image; new atoms need proper hash functions
+    #
+    # (*) Rational values are sympy Rationals which are always held in reduced form
+    # and thus are hashed correctly (see sympy/core/numbers.py:Rational.__eq__()).
+
+    def __init__(self):
+        self._hashes = defaultdict(list)
+
+    def selector(self):
+        hashes = self._hashes
+        return lambda elem: hashes[hash(elem)]
+
+    def same(self):
+        return lambda a, b: a.same(b)
+
+
+class _GatherBin:
+    def __init__(self, item):
+        self._items = [item]
+        self.add_to = self._items.append
+
+    def from_python(self):
+        return Expression('List', *self._items)
+
+
+class _TallyBin:
+    def __init__(self, item):
+        self._item = item
+        self._count = 1
+
+    def add_to(self, item):
+        self._count += 1
+
+    def from_python(self):
+        return Expression('List', self._item, Integer(self._count))
+
+
+class _DeleteDuplicatesBin:
+    def __init__(self, item):
+        self._item = item
+        self.add_to = lambda elem: None
+
+    def from_python(self):
+        return self._item
+
+
+class _GatherOperation(Builtin):
+    rules = {
+        '%(name)s[list_]': '%(name)s[list, SameQ]'
+    }
+
+    messages = {
+        'normal': 'Nonatomic expression expected at position `1` in `2`.',
+        'needlist': 'expecting a List as first parameter.',
+        'smtst': ("Application of the SameTest yielded `1`, which evaluates "
+                  "to `2`. The SameTest must evaluate to True or False at "
+                  "every pair of elements."),
+    }
+
+    def apply(self, list, test, evaluation):
+        '%(name)s[list_, test_]'
+
+        if list.is_atom():
+            expr = Expression(self.get_name(), list, test)
+            evaluation.message(self.get_name(), 'normal', 1, expr)
+            return
+
+        if list.get_head_name() != 'System`List':
+            evaluation.error(self.get_name(), 'needlist')
+            return Symbol('$Aborted')
+
+        if _is_sameq(test):
+            return self._gather(list, _FastEquivalence())
+        else:
+            return self._gather(list, _SlowEquivalence(test, evaluation, self.get_name()))
+
+    def _gather(self, a_list, equivalence):
+        bins = []
+        Bin = self._bin
+
+        select = equivalence.selector()
+        same = equivalence.same()
+
+        for elem in a_list.leaves:
+            selection = select(elem)
+            for prototype, add_to_bin in selection:  # find suitable bin
+                if same(prototype, elem):
+                    add_to_bin(elem)  # add to existing bin
+                    break
+            else:
+                a_bin = Bin(elem)  # create new bin
+                selection.append((elem, a_bin.add_to))
+                bins.append(a_bin)
```

Comment:
It should be `O(n^2)` only for the worst case, i.e. only if the hash function fails us, as it's really two nested levels of bins here: in the `_FastEquivalence` case, the data structure built in the code above is actually this: `selects = {hash1: [(prototype1, add_to_bin1)], hash2: [(prototype2, add_to_bin2)], ...}`, with `add_to_bin`s being nothing but appends to lists.

`select(elem)` calls `_FastEquivalence`'s `selector()` function which is `lambda elem: hashes[hash(elem)]`, with `hashes` being a `defaultdict(list)`.

So, `select(elem)` initially returns an empty list for an object. Later, when an object with the same hash comes along, a 1-element list with prototype as shown above is returned. So we might have `selects = {hash1: [(prototype1, add_to_bin1)]}` and an object with a hash of `hash1`, so `select(elem)` returns `(prototype1, add_to_bin1)`.

In this latter case, the for loop is entered, `same` is called (with the existing `prototype` and the new element). If `same` yields `true`, the element gets added to the nth bin, but no new prototype entry is added, i.e. we stay with `selects = {hash1: [(prototype1, add_to_bin1)]}`, only the bin behind `add_to_bin1` got larger one element.

Only if `same` yields false, will a second (or nth) prototype entry be added. Having multiple prototype entries in one hash bucket means that those entries share the same hash value but had `same` return `false` between them. This should be a border case with a proper hash function.

I think I should rename the `selector()` method to `select()` in order to match with the naming above.